vowpalwabbit/slates.cc

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "slates.h"
+#include "example.h"
+#include "slates_label.h"
+#include "reductions.h"
+#include "global_data.h"
+
+#include "vw.h"
+#include <algorithm>
+#include "cb_algs.h"
+#include "cb_adf.h"
+#include "conditional_contextual_bandit.h"
+#include "decision_scores.h"
+#include "action_score.h"
+#include "ccb_label.h"
+
+using namespace VW::config;
+
+namespace slates
+{
+template <bool is_learn>
+void slates_data::learn_or_predict(LEARNER::multi_learner& base, multi_ex& examples)
+{
+  _stashed_labels.reserve(examples.size());
+  for (auto& example : examples)
+  {
+    _stashed_labels.push_back(std::move(example->l.slates));
+  }
+
+  const auto num_slots = std::count_if(examples.begin(), examples.end(),
+      [](const example* example) { return example->l.conditional_contextual_bandit.type == CCB::example_type::slot; });
+
+  float global_cost = 0.f;
+  bool global_cost_found = false;
+  uint32_t action_index = 0;
+  size_t slot_index = 0;
+  std::vector<std::vector<uint32_t>> slot_action_pools(num_slots);
+  for (size_t i = 0; i < examples.size(); i++)
+  {
+    CCB::label ccb_label;
+    memset(&ccb_label, 0, sizeof(ccb_label));
+    CCB::ccb_label_parser.default_label(&ccb_label);
+    const auto& slates_label = _stashed_labels[i];
+    if (slates_label.type == slates::example_type::shared)
+    {
+      ccb_label.type = CCB::example_type::shared;
+      if (slates_label.labeled)
+      {
+        global_cost_found = true;
+        global_cost = slates_label.cost;
+      }
+    }
+    else if (slates_label.type == slates::example_type::action)
+    {
+      ccb_label.type = CCB::example_type::action;
+      slot_action_pools[slates_label.slot_id].push_back(action_index);
```

Comment:
slot_id is always [0, num_slots)?

Reply:
Yes, it is meant to be. But this might not be guaranteed...

Reply:
Adding a throw here to enforce this, until I get frontload this into parsers